stream-parent/stream-client/src/main/java/org/interledger/stream/sender/SimpleStreamSender.java

```diff
@@ -308,46 +419,75 @@ public SendMoneyAggregator(
             .data(streamPacketData)
             .build();
 
-        // Send it!
-        final CompletableFuture<InterledgerResponsePacket> sendMoneyFuture = CompletableFuture.supplyAsync(() -> {
-          congestionController.prepare(amountToSend);
-          return link.sendPacket(preparePacket);
-        }, executorService).thenApply(responsePacket -> {
-          // Executed in the same thread as above
-          responsePacket.handle(
-              (fulfillPacket -> {
-                handleFulfill(sequence.get(), amountToSend, fulfillPacket);
-              }),
-              (rejectPacket -> {
-                handleReject(sequence.get(), amountToSend, rejectPacket);
-              })
-          );
-          return responsePacket;
-        });
+        try {
+          // don't submit new tasks if the timeout was reached within this iteration of the while loop
+          if (canBeScheduled(timeoutReached.get(), streamPacket.sequence())) {
+            // call this before spinning off a task. failing to do so can create a condition in the
+            // soliderOn check where we will incorrectly evaluate hasInFlight on the congestion
+            // controller to not reflect what we've actually scheduled to run, resulting in the loop
+            // breaking prematurely
+            congestionController.prepare(amountToSend);
```

Comment:
We should think a bit more about this. While we want the `congestionController.prepare(amountToSend)` to be called outside the execution, I think we need to account for two cases:

* An error is thrown while calling `executorService.submit(()` (line `430`).
* An Exception is thrown and processed in line `442`.

In either case, it seems like we would want to unwind the call to `congestionController.prepare(amountToSend)` to avoid deadlock per that comment. However, it seems like the call to prepare in the congestion controller should be more tightly connected here.

Any thoughts?

Reply:
One thought is that the call to unwind a prepare be idempotent somehow, but that would require some kind of means of tracking a call to prepare by returning a uuid or something similar. Due to the fact that we're scheduling we can't rely on a try/finally approach since the finally executes outside of the thread if you're trying to guard against an exception on submit.